### ISSUE TITLE
Very poor geocoding performance for address where address is well formed

### ISSUE THREAD

**Issue by ColmDC** [DigitalCommons/se-open-data#1]
**Describe the bug**
A co-op appears in a completely incorrect location, despite the address seeming to be reasonably formed.

**To Reproduce**
1. Open 
2. Search for "Proper Tea Housing Co-operative Limited" 
3. Zoom in an you can see it is located at 

Their address on both the FCA and our Co-ops UK's CRM indicates the Registered address as 40 Hampton Grove, Belfast, BT7 3DG, but from what I can see on the map, it's showing up as the city hall in central Belfast!

**Expected behavior**
To appear near 40 Hampton Grove, Belfast, BT7 3DG

**Screenshots**
![image.png](https://images.zenhubusercontent.com/63665589a74876241a832285/fca784f6-e17e-4297-8e2f-ec3cc1f59522)

**Comment by ColmDC** [comment]
Tested address using https://apidocs.geoapify.com/playground/geocoding/

and it generated a sensible result, so it is probably not a geocoding issue.

**Comment by ColmDC** [comment]
It is positioning the marker consitently with what is in the standard.csv file, but it is clearly wrong.
There are several co-ops positioned there. Is this address a default address that is used when something else fails. I'm wondering if it is used when the confidence level returned for the address is too low?

**Comment by wu-lee** [comment]
Ok, it's not a default location. It's a cached location.  There's a big cache file which the sausage machine uses to speed up geolocation - otherwise it takes an eternity.  Thing is, the cache has no way to be cleared except by deleting or editing the cache file, a massive 35MB JSON monster.

So easy to fix, but it will take a while to regenerate the addresses.

**Comment by ColmDC** [comment]
Where does the cache live? Why is the cached location wrong?

**Comment by wu-lee** [comment]
The cache lives in a temp directory on the sausage machine. Ad hoc backups of this live in the `geocaches` project.

I don't know why it's wrong - I'm attempting to regenerate it and maybe I'll find out.

**Comment by wu-lee** [comment]
Ok, the regenerated data still has this problem.

Looking in the cache, there are two addresses for 40 Hampton Grove in Belfast.
- 40 Hampton Grove, Belfast, Belfast, BT7 3DG, United Kingdom, which resolves to:
  - lat/lng 54.596391,-5.9301829
  - confidence 0.02
  - Belfast, BT1 5JD, United Kingdom
- 40 Hampton Grove, Belfast, Belfast, BT7 3DG, Northern Ireland, which resolves to
  - lat/lng 54.5754252,-5.9287377
  - confidence 0.44
  - Hampton Grove, Down, BT7 3NA, United Kingdom

The incorrect one is the former, the correct the latter.  The former might to be the result of not having "Northern Ireland" in the address - but implies without that basically just resolves to the middle of Belfast (which seems a bit daft, but then it is low confidence.)

So maybe our geocoder is just taking the first listed match returned, and not ranking by confidence?  It's a bit opaque what it's doing, and my earlier attempts to unpack it have all been shelved.

But this highlights why small changes in the address can get jumps in location with the DotCoop data.

**Comment by ColmDC** [comment]
So good an bad news. Good that it looks like this bug could be a cause of other geocoding symptoms but bad, that if might need a lareger effort to decode the old geocoding code. Does the LX project use a geocoder?

**Comment by wu-lee** [comment]
Not sure - suspect they delegate everything to MapBox. John would be the person to ask. @King-Mob, do you know?

**Comment by wu-lee** [comment]
I've spent a bit of time trying to deduplicate the geocoding code in the sausage machine, with some success, using regression tests.

There are other issues, such as changing CSV schema fields resulting in Country Name -> Country ID resulting in blank countries in addresses, in turn resulting in inaccurate geocoding (geoapify defaults to the US when no country is given it seems)

**Comment by King-Mob** [comment]
We use the mapbox geocoding service. It's alright. Never had cause to verify whether the results are accurate on a house number level, though

**Comment by wu-lee** [comment]
I'm trying to get the Geoapify service to return more than one location for this problem address.  It doesn't seem to want to - so that's the end of the idea that we can select the best confidence match - there's only one, even when I vary the address.

The address that the sausage machine is using: "40 Hampton Grove, Belfast, Belfast, BT7 3DG, United Kingdom"

So these all get one result except where stated:
- 40 Hampton Grove, Belfast, Belfast, BT7 3DG, United Kingdom: *wrong location, central Belfast*
- 40 Hampton Grove, Belfast, Belfast, BT7 3DG: *correct location*
- 40 Hampton Grove, Belfast, Belfast, BT7 3DG, Northern Ireland: *correct location*
- 40 Hampton Grove, Belfast, BT7 3DG, Great Britain: *correct location*
- 40 Hampton Grove, Belfast, BT7 3DG, United Kingdom: *wrong location, central Belfast*
- BT7 3DG, United Kingdom: *wildly wrong, several locations across UK.*
- Belfast, BT7 3DG, United Kingdom: *wrong, several locations around Belfast*
- Belfast, BT7 3DG: *wrong, somewhere in Belfast and South Africa*
- 40 Hampton Grove, Belfast, BT7 3DG, UK: *wrong location, central Belfast*
- 40 Hampton Grove, Belfast, BT7 3DG, GB: *wrong location, central Belfast*
- Belfast, BT7 3DG, UK: *wrong, somewhere near Birmingham*

So clearly here Geoapify thinks that Belfast is not in the United Kingdom. It thinks it is in Great Britain, but not GB or UK. (Although most of these results seem to be sourced as from OpenStreetMap, so perhaps that's at fault.)  

It does better with the Jersey address I tried: Co-operative House 57 Don Street, St Helier, JE2 4TR, Jersey. It copes better with shortened versions, and will cope with Great Britain instead of Jersey - but not GB.

It also does better with other NI addresses, for example: 32 Main St, Ballykelly, Causeway Coast and Glens, BT49 9HS, United Kingdom. This has high confidence and looks right. BUT - if you remove the "United Kingdom" or replace it with "Great Britain", it resolves to: 32 Main Street Andes, NY 13731, United States of America.  Wildly wrong! And with a 63% confidence.

For the Co-ops UK standard.csv, we currently record the country as United Kingdom or GB, depending on whether the version of the code is using the older schema - using Country Name - or the newer - using Country ID.  Although when geocoding, the address isn't (wasn't) used at all, just the postcode.

I was thinking we could try to supply Great Britain as the country for all  the data if we use Geoapify... But that last case, the one resolving it to the US, blew that idea out of the water.  

I thought of trying to just geocode using the postcode and country code ( which would be GB) using the API call used in the Geoapify example here:

https://apidocs.geoapify.com/docs/geocoding/forward-geocoding/#geocode-zipcodes

Unfortunately, this doesn't work. Worryingly, their example of looking up a UK postcode doesn't:

This: https://api.geoapify.com/v1/geocode/search?text=B27&lang=en&limit=10&type=postcode&filter=countrycode:gb&&format=json&apiKey=d548c5ed24604be6a9dd0d989631f783
Gets:
```
{
  "results": [],
  "query": {
    "text": "b27",
    "parsed": {
      "postcode": "b27",
      "expected_type": "unknown"
    }
  }
}
```
As does the postcode in question, BT7 3DG:

https://api.geoapify.com/v1/geocode/search?text=BT73DG&lang=en&limit=10&type=postcode&filter=countrycode:gb&&format=json&apiKey=d548c5ed24604be6a9dd0d989631f783

Gets:
```
{
  "results": [],
  "query": {
    "text": "bt7 3dg",
    "parsed": {
      "postcode": "bt7 3dg",
      "expected_type": "unknown"
    }
  }
}
```

Postcodes from other parts of the UK do work, however, e.g. OX29 7SA

It did occur to me that the Belfast postcode is incorrect or malformed. But I've checked it is correct by using this website, and it seems to be: https://www.doogal.co.uk/BatchGeocoding

So: so far, no simple solution which can geocode all the cases I've checked.

**Comment by ColmDC** [comment]
How big a job is changing to a different geocoding service?
Do we know what percentage of queries produce more than on option?

**Comment by wu-lee** [comment]
Changing to a different service: maybe order of a few hours now I have a regression test and have deduplicated/tidied some of the code.

Percentage: I don't know as the current geocoder doesn't record that.

Here's a quick attempt to chart the frequency of the confidence levels of all the currently cached geocoded addresses, however.

There are a significant number below 30%, and even 10%.

```
ruby -rjson -e 'f=Array.new(101, 0); t=0; JSON.load(File.read(ARGV[0])).each {|k,v| c= v.dig("rank","confidence").to_f*100; f[c.to_i]+= 1;t+=1;} ; f.each_with_index {|f,b| p = (f.to_f*100/t).to_i; puts "#{"% 4d" % b}%: #{"% 3d" % p}% (#{"% 5d" % f}) #{"*"*(p+1.0)}"}; puts "total: #{t}"' ../open-data/caches/geodata_cache.json
   0%:   2% (  392) ***
   1%:   0% (    7) *
   2%:   0% (   99) *
   3%:   0% (    7) *
   4%:   0% (   32) *
   5%:   0% (   16) *
   6%:   0% (   33) *
   7%:   0% (   16) *
   8%:   0% (    9) *
   9%:   0% (   21) *
  10%:   0% (    8) *
  11%:   0% (   15) *
  12%:   0% (  180) *
  13%:   0% (    8) *
  14%:   0% (   25) *
  15%:   0% (   35) *
  16%:   0% (    5) *
  17%:   0% (   18) *
  18%:   0% (   27) *
  19%:   0% (   26) *
  20%:   0% (   18) *
  21%:   0% (    4) *
  22%:   0% (   27) *
  23%:   0% (    5) *
  24%:   0% (   50) *
  25%:   6% ( 1179) *******
  26%:   0% (    2) *
  27%:   0% (   20) *
  28%:   0% (   17) *
  29%:   0% (    3) *
  30%:   0% (   23) *
  31%:   0% (   27) *
  32%:   0% (    3) *
  33%:   0% (    6) *
  34%:   0% (    7) *
  35%:   0% (   43) *
  36%:   0% (    8) *
  37%:   0% (   11) *
  38%:   0% (    2) *
  39%:   0% (   61) *
  40%:   0% (   39) *
  41%:   0% (    3) *
  42%:   0% (    8) *
  43%:   0% (    2) *
  44%:   2% (  550) ***
  45%:   0% (  174) *
  46%:   0% (    4) *
  47%:   0% (    9) *
  48%:   1% (  310) **
  49%:   0% (    1) *
  50%:   3% (  654) ****
  51%:   0% (  135) *
  52%:   0% (    3) *
  53%:   0% (   11) *
  54%:   0% (    6) *
  55%:   0% (    4) *
  56%:   0% (  162) *
  57%:   0% (   11) *
  58%:   0% (    2) *
  59%:   0% (    7) *
  60%:  11% ( 2129) ************
  61%:   0% (    2) *
  62%:   0% (    7) *
  63%:  11% ( 2233) ************
  64%:   0% (    2) *
  65%:   0% (    8) *
  66%:   0% (  151) *
  67%:   0% (    7) *
  68%:   0% (    3) *
  69%:   0% (    1) *
  70%:  24% ( 4632) *************************
  71%:   0% (    3) *
  72%:   0% (   38) *
  73%:   0% (    3) *
  74%:   0% (    2) *
  75%:   0% (   24) *
  76%:   0% (    5) *
  77%:   0% (    4) *
  78%:   0% (    1) *
  79%:   0% (    1) *
  80%:   0% (    9) *
  81%:   0% (  120) *
  82%:   0% (    5) *
  83%:   0% (    0) *
  84%:   0% (    0) *
  85%:   0% (   47) *
  86%:   0% (    4) *
  87%:   0% (    2) *
  88%:   0% (    0) *
  89%:   0% (    0) *
  90%:   3% (  634) ****
  91%:   0% (    0) *
  92%:   0% (    0) *
  93%:   0% (    0) *
  94%:   0% (    1) *
  95%:   0% (  147) *
  96%:   0% (    0) *
  97%:   0% (    0) *
  98%:   0% (    0) *
  99%:   0% (    0) *
 100%:  21% ( 4069) **********************
total: 18884
```

**Comment by ColmDC** [comment]
Are you awaiting go ahead to try a different geocoder @wu-lee?
Go for it.

**Comment by King-Mob** [comment]
@wu-lee message on element demonstrating mapbox geocoding request

**Comment by ColmDC** [comment]
Dropped from this Epic, but still needed, so keep in NextUp

**Comment by wu-lee** [comment]
The current code for geocoding in the se-open-data library does a CSV-at-once, which means it has to be done in a separate, subsequent step from the conversion. It also means the precise mapping of fields to addresses and back to fields again is somewhat hardwired and difficult to alter. Plus the details of the operation are quite (perhaps unnecessarily?) complex, so hard to change without fear of breaking things.

So a good way forward seems to be to implement some geocoding solution which operates on a single case, which can then be invoked from within the field mapping process at the start of the conversion. A possible disadvantage might be that doing it in the initial field mapping loop then can't make a batch conversion call on a long list of addresses, which some APIs support. However, I don't think this is what the current GeoAPIfy  code does, so no loss there, plus if it were really needed, a second loop could be implemented in the converter script, which wouldn't be much harder than now, since a second step is currently mandatory.

Furthermore, there seems to exist a mature geocoding Ruby gem, called simply [geocoder][1], which looks like it does all we need and more:
- supports GeoAPIfy
- supports MapBox
- and others
- supports caching (including custom caching methods)
- simple API: `Geocoder.search("some address")`

The se-open-data library could facilitate all this with some convenience functions, in particular to support caching in a similar way as it has been (a simple JSON file), or (if needed) support for batch geocoding.

[1]: https://www.rubygeocoder.com/

**Comment by ColmDC** [comment]
Looks like a great option.

**Comment by ColmDC** [comment]
Do you want to manage the work to implement in this ticket? @wu-lee or does it deserve 1 or more new issues?

**Comment by wu-lee** [comment]
I think I can just implement it in one issue. Mostly done already, just tidying for reuse, and adding some nominal tests.

**Comment by ColmDC** [comment]
Great. Let's see if we can also close or update these Issues when we complete this one.

https://github.com/DigitalCommons/technology-and-infrastructure/issues/84
https://github.com/DigitalCommons/open-data-and-maps-outreach/issues/231
https://github.com/DigitalCommons/dotcoop-project/issues/62

**Comment by ColmDC** [comment]
I guess we need to decide on a threshold confidence below which we treat the geolocation as a fail, and don't generate for the entry and treat it as if there was no address. Or perhaps we need 3 states. No address, address fails geocoding and success. The two former have their dialogs positioned in the default location, but something in the dialog distinguished no addresses from failed addresses? 

If the confidence level is recorded in the csv file, we may not need to generate a seperate quality report.

**Comment by wu-lee** [comment]
Ok, so new release v2.2.0 of se-open-data gem has the new slimline geocoding functions (which are a thin shim around `geocoder`, designed to allow tweaking for our convenience).

Simultaneously, all the old geocoding code has been removed, along with a lot of CSV cleaning and manipulating routines only used by really ancient sausage-machine scripts using Make. If I've deleted anything important, will discover later, but I don't think so.

Hence the release has gone from v2.1.11 -> v2.2.0

The dotcoop dataset has been updated to use this version. Some other things have been fixed too (specifically, since se-open-data is a bit hardwired, it needs `USING_ICA_ACTIVITIES = true` set in the config to switch from the activities-modified default - yes, the selection of vocabs is fairly fixed).

Also, as a test, the limsurvey datasets have been updated to use it as well. This is a jump from the 1.x series of se-open-data to the 2.2.x series.  They seem to work, although an expectation for the countries-iso vocab needed the v2.1 ESSGLOBAL vocabs to be re-deployed on dev.vocabs.solidarityeconomy.coop to provide that. (production version done too for good measure).

So dotcoop still uses the geopaify service, but has been adjusted to record the confidence returned from it, in a field "Geo Container Confidence".  See the uploaded CSV, when if finally rebuilds (it's taking some time):

https://dev.data.solidarityeconomy.coop/dotcoop/standard.csv

The limesurvey dataset converter hasn't been adjusted to record this, so their CSVs have not changed.

**Comment by wu-lee** [comment]
> Great. Let's see if we can also close or update these Issues when we complete this one.
> 
> [DigitalCommons/technology-and-infrastructure#84](https://github.com/DigitalCommons/technology-and-infrastructure/issues/84)

I think that one is pretty sparse, and can be closed. The geocoding has been reviewed and replaced.

> [DigitalCommons/open-data-and-maps-outreach#231](https://github.com/DigitalCommons/open-data-and-maps-outreach/issues/231)

This is ancient and is also closable. In terms of improving the geocoding, I think we could look at the confidence data we get now, but also compare and contrast geoapify and mapbox (possibly google too.) In order to see how much variance there is. Where it's not working well some tweaking may be required, but first we have to spot where it's not working well.

I do notice that DotCoop's address data isn't very well cleaned up, incidentally. Inclusion of names of people/organisations in with the addresses, for example, and spurious punctuation which looks like empty fields concatenated with commas, resulting in too many commas.

> [DigitalCommons/dotcoop-project#62](https://github.com/DigitalCommons/dotcoop-project/issues/62)

This is now implemented, at least for the dotcoop data. Other dataset converters will need to be modified to a) use the new geocoding stuff, b) insert the confidence field into the CSV, on a case-by-case basis.

**Comment by wu-lee** [comment]
> I guess we need to decide on a threshold confidence below which we treat the geolocation as a fail, and don't generate for the entry and treat it as if there was no address. Or perhaps we need 3 states. No address, address fails geocoding and success. The two former have their dialogs positioned in the default location, but something in the dialog distinguished no addresses from failed addresses?
> 
> If the confidence level is recorded in the csv file, we may not need to generate a seperate quality report.

The confidence is recorded, for dotcoop, and can be in other cases if they are amended as mentioned in the last comment.

The confidence field will be empty if the geocoding fails. Otherwise there'll be a number 0-100, with 100 being the best.  No address will be clear when the address fields are empty. So this is amenable to being investigated by anyone with the URL to the CSV file and a spreadsheet.

**Comment by wu-lee** [comment]
Dotcoop data now regenerated with confidences. Stats:

```
	Unique values:         3
	Smallest value:        0
	Largest value:         100
	Sum:                   219,000
	Mean:                  50.636
	Median:                100
	StDev:                 50.002
	Most common values:    100 (2190x)
	                       0 (2135x)
	                       None (45x)
``` 

So there are a lot of zeros...

**Comment by wu-lee** [comment]
Ok, found a bug, which flattens the confidence to 0 or 100. Fixed.

As an aside: in the process of trying to show a frequency table of confidence values, discovered the Polars-df and Vega gems, which provide Pandas-like functionality for Ruby. IOW this:

```
gem install polar-df # dataframes slicing and dicing
gem install vega # plotting
gem install sculpt # adds HTML rendering needed to wrap the raw vega-generated plot HTML

ruby -rpolars-df -rvega -rsculpt -e 'plot = Polars.read_csv("generated-data/standard.csv")[0..,23].sort.value_counts.plot.to_s; Sculpt.render_doc { head { %w(vega@5 vega-lite@5 vega-embed@6).each{|s| js "https://cdn.jsdelivr.net/npm/"+s} }; body  plot }' >plot.html
```

Can generate this frequency graph of the Co-ops UK confidences, like this:

![image](https://github.com/DigitalCommons/se-open-data/assets/67825/f95030e6-da67-4283-9f9a-553e14e1778b)

And for DotCoop:

![image](https://github.com/DigitalCommons/se-open-data/assets/67825/dbcb68f8-c1da-43d0-90ab-bcc08b7c6e5f)


These tools look useful more generally for CSV file processing and interpretation. Polars-df is backed by a fast Rust implemented data-frames library, so is likely to be much faster than anything written in Ruby.

Anyway. Those graphs show there are significant amounts of low-confidence results, but maybe not as bad as the buggy answer!

Updated Co-ops UK data stats from csvstat:
```
	Unique values:         74
	Smallest value:        0
	Largest value:         100
	Sum:                   489,627
	Mean:                  66.453
	Median:                70
	StDev:                 26.725
	Most common values:    100 (1771x)
	                       70 (1683x)
	                       63 (1112x)
	                       25 (518x)
	                       50 (307x)
```

And for DotCoop:
```
	Unique values:         87
	Smallest value:        0
	Largest value:         100
	Sum:                   312,423
	Mean:                  73.408
	Median:                100
	StDev:                 33.749
	Most common values:    100 (2164x)
	                       25 (421x)
	                       50 (340x)
	                       90 (293x)
	                       0 (222x)
```

**Comment by wu-lee** [comment]
I've now added a "Geocoded Address" field to both the Co-ops UK and DotCoop output CSVs so that when the confidence looks bad, the address can checked directly against the relevant geocoder.

So currently, that means the csv files here:
- https://dev.data.solidarityeconomy.coop/coops-uk/standard.csv
- https://dev.data.solidarityeconomy.coop/dotcoop/standard.csv

Have fields (or will when the generator finishes) which can be pasted into the GeoAPIfy playground page to see what results come back, and try and determine why it's bad (if it is).

https://apidocs.geoapify.com/playground/geocoding/

I've tried a few. Seems to be useful. It's not always easy to see why it's bad. For instance, "Slackadale, Turrif, Aberdeenshire, AB53 5RL, Scotland" has a confidence of 0, but if you strip off the first to parts, "Aberdeenshire, AB53 5RL, Scotland" has a confidence of 100%. Less is more.

**Comment by wu-lee** [comment]
Ok, I've re-run the conversions with the mapbox geocoder for DotCoop and Co-ops UK.

These are the confidence frequency tables for Co-ops UK:

![image](https://github.com/DigitalCommons/se-open-data/assets/67825/22810c93-db31-4a7f-a0ec-479fc74433a0)

And DotCoop:

![image](https://github.com/DigitalCommons/se-open-data/assets/67825/b51a6ec8-63d4-4cab-8e0f-200ee6cebb9a)

Note that Mapbox confidence values are obtained from their "relevance" result which looks to be a slightly different thing:

https://docs.mapbox.com/help/getting-started/geocoding/#textual-relevance

The mapbox playground for testing the geocoding service is here:

https://docs.mapbox.com/playground/geocoding/

**Comment by wu-lee** [comment]
Checking the Proper Tea Housing Co-op address above. The geocoded address in both cases (GeoAPIfy and Mapbox) is now "40 Hampton Grove, Belfast, Belfast, BT7 3DG, Northern Ireland"

GeoAPIfy:
- confidence: 44%
![image](https://github.com/DigitalCommons/se-open-data/assets/67825/11e8aa28-848a-4386-aa17-b0edb1c7d8a8)

There are actually two matches very close to each other with 44%. Both look plasibly correct.


Mapbox:
- confidence (AKA textual relevance): 79%
![image](https://github.com/DigitalCommons/se-open-data/assets/67825/698adedc-ba39-412b-b17c-8acf9fe8606c)

This matches the second of the plausibly correct GeoAPIfy result.

So overall, for this case, even the GeoAPIfy result is better. Mainly the result of a cleaner address string, I think - part of the refactor allows dataset converters to construct their own address strings. The address used here differs from the failing one above in that it uses "Northern Ireland" instead of "United Kingdom".

The new Geoapify data is now deployed for Co-ops UK and used here:
- https://dev.coopsuk.solidarityeconomy.coop/

Compare with the old data used here:
- https://coopsuk.solidarityeconomy.coop/

The DotCoop dev map has the new data, however the production map is still waiting for a URI from Johnathan (see https://github.com/DigitalCommons/dotcoop-project/issues/60)

**Comment by wu-lee** [comment]
Checking island outliers for Co-ops UK. Jersey has two organisations, as before. Isles of Scilly, one, as before. These look correct.

On the other hand, I see for the updated GeoAPIfy data, that there are three organisations shown on the Isle of Man. One, Manx Credit Union, looks like it is correct (confidence 100%) - the other two should be on the Scottish isles of Harris (35%) and North Uist (35%).

And the new data has organisations in the Netherlands and Catalonia, both of which are wrong, and should be in Kilsby (14%) and the Isle of Barra (0%), respectively.  Zooming out I see dozens more outside the UK.

An example: "29 Gensing Road, St Leaonards On Sea, East Sussex, TN38 0HE, England" geocodes to a location in Indonesia, with a confidence of 12%.  Pasting this into the GeoAPIfy playground, shows there are two matches in Indonesia

GeoAPIfy results:
| Identifier | Name                                           | Geo Container Confidence | Geocoded Address                                                                         | Geo Container Latitude | Geo Container Longitude |
| ---------- | ---------------------------------------------- | ------------------------ | ---------------------------------------------------------------------------------------- | ---------------------- | ----------------------- |
| R001631    | Harris Community Co-operative                  |                       35 | The Clachan Leverburgh, Isle of Harris, Eilean Siar, HS5 3TS, Scotland                   |                54.157… |                 -4.566… |
| R003219    | North Uist and Benbecula Livestock (Marketing) |                       35 | Estate Office Lochmaddy, Isle of North Uist, Eilean Siar, HS6 5AA, Scotland              |                54.169… |                 -4.523… |
| R009259    | Buth Bharraigh                                 |                        0 | Old Co-op Building Castle Bay Isle of Barra, Eilean Siar, Eilean Siar, HS9 5XD, Scotland |                57.358… |                 -5.559… |
| R010002    | Gensing Workers' Co-operative                  |                       12 | 29 Gensing Road, St Leaonards On Sea, East Sussex, TN38 0HE, England                     |                -4.054… |                121.590… |
| R015178    | Manx Credit Union                              |                      100 | Ragnall House 18 Peel Road, Douglas, , IM1 4LZ, Isle of Man                              |                54.149… |                 -4.487… |
| R022403    | Kilsby Village shop                            |                       14 | 5 Emery Row Malt Mill Green, Kilsby, West Northamptonshire, CV23 8YY, England            |                52.378… |                  4.861… |

MapBox results:
| Identifier | Name                                           | Geo Container Confidence | Geocoded Address                                                                         | Geo Container Latitude | Geo Container Longitude |
| ---------- | ---------------------------------------------- | ------------------------ | ---------------------------------------------------------------------------------------- | ---------------------- | ----------------------- |
| R001631    | Harris Community Co-operative                  |                    0 | The Clachan Leverburgh, Isle of Harris, Eilean Siar, HS5 3TS, Scotland                   |                57.771… |                 -7.015… |
| R003219    | North Uist and Benbecula Livestock (Marketing) |                    0 | Estate Office Lochmaddy, Isle of North Uist, Eilean Siar, HS6 5AA, Scotland              |                57.600… |                 -7.161… |
| R009259    | Buth Bharraigh                                 |                    0 | Old Co-op Building Castle Bay Isle of Barra, Eilean Siar, Eilean Siar, HS9 5XD, Scotland |                56.956… |                 -7.488… |
| R010002    | Gensing Workers' Co-operative                  |                    0 | 29 Gensing Road, St Leaonards On Sea, East Sussex, TN38 0HE, England                     |                50.853… |                  0.558… |
| R015178    | Manx Credit Union                              |                    0 | Ragnall House 18 Peel Road, Douglas, , IM1 4LZ, Isle of Man                              |                54.150… |                 -4.478… |
| R022403    | Kilsby Village shop                            |                    0 | 5 Emery Row Malt Mill Green, Kilsby, West Northamptonshire, CV23 8YY, England            |                52.334… |                 -1.175… |

So clearly using the GeoAPIfy confidence score and some address clean up alone is not enough. The new results from the new  are better in places, but way worse in others.

**Comment by wu-lee** [comment]
To illustrate the differences between the GeoAPIfy prior to the changes in the geocoding mechanism, and after, and with MapBox (also after), here are three Co-ops UK maps, one for each.

- https://dev.maps.solidarityeconomy.coop/temp/coopsuk-geoapify-old/
- https://dev.maps.solidarityeconomy.coop/temp/coopsuk-geoapify-new/
- https://dev.maps.solidarityeconomy.coop/temp/coopsuk-mapbox-new/

**Comment by wu-lee** [comment]
Note that the first map's dataset does not have geocoded address or confidence fields, so they appear in the popup as "N/A"

**Comment by ColmDC** [comment]
We also can't always trust the confidence levels, as I've seen examples of 0% confidence (NTFC Supporters Trust on GeoApify) where the match is pretty close, closer in fact than that 81%! from Mapbox.

**Comment by ColmDC** [comment]
So really useful results, but what is next? What are your recommendations @wu-lee ?

**Comment by wu-lee** [comment]
Well, I'm not totally sure, but it might look like we should use MapBox rather than GeoAPIfy, at least for CUK. Reasoning: all the pins are in the UK, which is a good sign.

I'm not really sure how to identify bad geolocations without using eyeballs.  I think perhaps we should delegate that to DotCoop? (Or their registrants)  But then if we have a list of bad cases, we might be able to find a way to tweak the queries to ensure better quality results, or otherwise conclude the geolocation can't be trusted full stop.  I suspect it won't be a simple rule, it'll be a quite labour-intensive process of discovering what works and what doesn't, and there might be more than one workaround which need to be applied in different cases.

Another tactic, which can be applied anyway, so to publish a disclaimer that geocoded locations are not guaranteed to be high quality, but we can't do much about that without a) decent and well formatted addresses and/or b) actual location data, so please check and supply us with those if the location is wrong.

Perhaps we can come up with some heuristics for badly-formed addresses. Like being super short, containing strange punctuation like @ symbols, lacking a known country, lacking a postcode, etc.

**Comment by ColmDC** [comment]
Okay, so for now shift to using Mapbox for Co-ops UK and stay with GeopApify for DotCoop. Let's close this ticket then, as we've made some good progress and the problem candiates that triggered it are much better. Can you present your observations at the next All Hands and we can have a broader discussion of Geocoding strategies?

**Comment by wu-lee** [comment]
Ok- coops-uk dataset has been rebuilt for development and redeployed here:

https://dev.data.solidarityeconomy.coop/coops-uk/

The map here will be using it as a matter of course:

https://dev.coopsuk.solidarityeconomy.coop/

It took literally hours for the RDF to build fully....

**Comment by ColmDC** [comment]
> It took literally hours for the RDF to build fully....

Luckily we only need to do it every quarter

**Comment by ColmDC** [comment]
> Luckily we only need to do it every quarter

Or it's monthly now. Is this the latest data?

**Comment by ColmDC** [comment]
> Or it's monthly now. Is this the latest data?

About box says q2 2023. That may not be true. However, if you haven't upgraded to latest, don't do it yet, as it deserves a fresh ticket and I'll include a few other tweaks and then go back to Co-ops UK with it.

**Comment by ColmDC** [comment]
Looking at the data, there are a lot more orgs in the data with geo locations than appear on the map. What criteria did you use to decide it they are displayed or not?

**Comment by wu-lee** [comment]
> > Luckily we only need to do it every quarter
> 
> Or it's monthly now. Is this the latest data?

I've not downloaded a new dataset, so probably not if it's now monthly.

**Comment by ColmDC** [comment]
> Looking at the data, there are a lot more orgs in the data with geo locations than appear on the map. What criteria did you use to decide it they are displayed or not?

Just checking you hadn't missed that question?

**Comment by wu-lee** [comment]
I remember replying to this, but the reply isn't here, so something must have gone wrong.

To answer: I've not changed  what goes on the map, or not intentionally. 

What used to happen: DotCoop data is converted to CSV, then in a following pass has geocoded locations added using the standard address fields, into the fields "Geo Coded Longitude" and "Geo Coded Latitude".  "Latitude" and "Longitiude" fields are empty, because no locations supplied by DotCoop.

What now happens: DotCoop data is converted to CSV. and in the same pass has geocoded locations added using information sourced (and cleaned up) from address fields selected by the converter. Again, these go into the fields "Geo Coded Longitude" and "Geo Coded Latitude".  "Latitude" and "Longitiude" fields remain empty, because no locations supplied by DotCoop.

In both cases, Mykomap will show a pin if one or other lat/lng field pairs are set.  It uses the geocoded ones only if the other ones are empty. If both are empty, the organisation is listed, but has no pin.

**Comment by ColmDC** [comment]
> To answer: I've not changed what goes on the map, or not intentionally.

I think this is the same issue that Lance noticed and the prod version of dotcoop for some reason has gone back to pointing at the old data, or something been copied over something in error. I thinks calling everything standard.csv and having no timestamps on the csvs is making this harder to debug.

**Comment by wu-lee** [comment]
There is a timestamp - the one on the file itself isn't easy to check because there isn't a directory index, and clicking the file itself downloads the CSV.

But the data factory creates a file with the timestamp the dataset was created (!= uploaded), which you can view if you substitute `meta.json` for `standard.csv`. (It's a diagnostic, so not actively publicised.)

Currently, https://data.solidarityeconomy.coop/dotcoop/meta.json has:

```
{
  "timestamp": "2023-11-09 14:50:40 UTC",
  "commit_id": "3cb92615cdbb3bcc151bbde9a86fcab9fe662449",
  "standard_data_hash": "b51d1d7adfe19e2be865427d83fc9c697bbb6cb7",
  "doc_hash": "2c3200220100304589e18917f061b4cdd9a57dd9",
  "user": "seopendata",
  "host": "prod-1"
}
```
So it's generated recently, from yesterday.  Using a commit ID starting with [3cb9261][1], which is also as expected. (It's the current `prod-1` branch; it has its own branch distinct from `master` for convenience, usually to allow hot-fixes)

So something else is up. Investigaging back in https://github.com/DigitalCommons/dotcoop-project/issues/60#issuecomment-1806002895

[1]: https://github.com/DigitalCommons/open-data/commit/3cb92615cdbb3bcc151bbde9a86fcab9fe662449

**Comment by wu-lee** [comment]
Ok, that last post was looking at the production data.


Currently, https://dev.data.solidarityeconomy.coop/dotcoop/meta.json has:

```
{
  "timestamp": "2023-11-07 13:44:56 UTC",
  "commit_id": "58a9c51b46f6051b827cd091f77d44aa5cb519ab",
  "standard_data_hash": "e841f71c38616f29430b6dc77039cae9148e0b27",
  "doc_hash": "96e7741c19f352c61dc76059e502f5038bbb700f",
  "user": "seopendata",
  "host": "dev-1"
}
```
So it's recent but a little older. This is because the geocoding stuff on development was added before the final steps getting `dotcoop2023` -> `dotcoop` on production, and the data reflects the time I manually prodded the data to be rebuilt. (Otherwise they should rebuild at the same time, when the data feed's download changes.)
